### PR TITLE
Fixed build process and added workflows to test dev and prod builds

### LINKED ISSUE #2730: API: Extremely Poor Docker Resource Utilization Efficiency
**Describe the bug**

We run a demonstration instance of Talawa-API on a GoDaddy VPS server running Ubuntu. It has the following resources:

1. 1 core
1. 2 GB of RAM
1. 40 GB of disk

Other information:

1. The demo instance is intended to create an evaluation environment for new GtiHub contributors and users alike as they decide to use Talawa. The DB of the demo instance gets reset every day.
2. Talawa API runs natively on this VPS server with acceptable performance with one user. The load average is approximately 1, which is the target value for a system with only 1 core.
3. When Talawa API runs on the server using docker. The load average reaches 130, the swap process is the top CPU resource user. The system is so overloaded that only one ssh session at a time is achievable.

The purpose of this issue is to find ways to tune all Talawa-API Dockerfile and app configurations to lower its CPU and RAM utilization by at least 75% 

1. With the current Docker performance very few developers or end users will want to try Talawa themselves.
2. This has been a recurring issue with Talawa API. The poor performance threatens the success of our current MongoDB based MVP.

**To Reproduce**

1. Run Talawa-API on a system
2. See excessive resource utilization

**Expected behavior**

1. Acceptable usage information such that it can run easily on a mid-range laptop without impacting its performance

**Actual behavior**

1. Poor performance

**Screenshots**

image

**Additional details**

**Potential internship candidates**

Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by adithyanotfound**
**Issue Number:**

Partially Fixes #2730 

**Did you add tests for your changes?**

Not required.

**Snapshots/Videos:**

https://youtu.be/xgz-WZik9Gg

https://github.com/user-attachments/assets/0e65fbd8-55f5-4612-93a2-0b300ef37f95

**If relevant, did you update the documentation?**

No

**Summary**

1. Fixes build process which was broken due to conficts between ES module and commonjs.
2. Added workflow to test both development and production builds 

**Does this PR introduce a breaking change?**

No

**Other information**

1. The prod build requires key.pem and cert.pem files (ssl). They build will fail without them.
2. Production does not support apollo sandbox due to security concerns and  being resource intensive. We can use local tools like postman to test in production.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

- **New Features**
	- Introduced a new job `Test-Builds` in the GitHub Actions workflow for testing development and production builds.
	- Added "AppUserProfile" to the `collectionName` enum in the sample data interface.

- **Bug Fixes**
	- Corrected the spelling of `persist-credentials` in the GitHub Actions workflow.
	- Fixed the index creation for `hourHistory.date` in the `EventVolunteer` model.
	- Corrected the misspelling of `advertisements` in the `Organization` schema.

- **Documentation**
	- Enhanced comments for clarity in various files, including server configuration and CORS handling.

- **Refactor**
	- Consolidated Mongoose imports across multiple model files for improved clarity and consistency.
	- Restructured the `Test-Application` job to depend on the new `Test-Builds` job.
	- Updated path resolution methods in the `uploadEncodedImage` and `uploadEncodedVideo` functions to accommodate ES module syntax.
	- Improved validation checks for logging properties in the `dbLogger` utility.

- **Tests**
	- Enhanced test suites for `uploadEncodedImage` and `uploadEncodedVideo` to include mocks for file system interactions and new test cases for directory creation.
	- Updated variable declarations in test files to use `const` for immutability.

### CODE REVIEW

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -296,6 +273,98 @@ jobs:
...
+        run: |
+          echo "Cleaning up background processes..."
+          pkill -f "npm run dev" || echo "No dev process found."
+          pkill -f "npm run prod" || echo "No prod process found."
+
+      - name: Delete key.pem and cert.pem
+        run: |
+          echo "Deleting key.pem and cert.pem..."
+          rm -f key.pem cert.pem
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply .github/workflows/pull-request.yml:367]: We don't need secure keys in workflow

**Comment by palisadoes** [comment]
Does this mean that we need to update the installation script for a production server to create the SSL certs?

**Comment by adithyanotfound** [comment]
> Does this mean that we need to update the installation script for a production server to create the SSL certs?

Yes. Currently `npm run generate:ssl-private-key` only generates key.pem. We also require it to generate cert.pem.

**Comment by palisadoes** [comment]
> > Does this mean that we need to update the installation script for a production server to create the SSL certs?
> 
> Yes. Currently `npm run generate:ssl-private-key` only generates key.pem. We also require it to generate cert.pem.

Please update the setup script and add the certs to `.gitignore`. Ideally they should be placed in a separate `certs/` directory that is empty in the repo.

We eventually need a single solution for production and development. There are too many options and cases to test. 

This has been broken for too long because nobody was using it.

**Comment by palisadoes** [comment]
Did you get the GraphQL endpoint to work?

**Comment by palisadoes** [comment]
Please add the required tests for the patch

**Comment by adithyanotfound** [comment]
> Did you get the GraphQL endpoint to work?

Yes

**Comment by vasujain275** [comment]
@adithyanotfound Great work on this, just add the required tests and this will be good to go in my opinion. We can then move on to deployment scripts and docker files. This patch alone will fix our docker performance issues.

File: .github/workflows/pull-request.yml [hunk 2]
```diff
@@ -255,46 +255,167 @@ jobs:
...
-          verbose: true
-          token: ${{ secrets.CODECOV_TOKEN }}
-          fail_ci_if_error: false
-          name: '${{env.CODECOV_UNIQUE_NAME}}'
+      - name: Clean up Background Processes
+        if: always()
+        run: |
+          echo "Cleaning up background processes..."
+          pkill -f "npm run dev" || echo "No dev process found."
+          pkill -f "npm run prod" || echo "No prod process found."
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply .github/workflows/pull-request.yml:412]: not required

File: .github/workflows/pull-request.yml [hunk 3]
```diff
@@ -296,6 +273,133 @@
...
+          if curl -f -k 'https://localhost:4001/graphql?query=%7B__typename%7D' \
+            -H 'Apollo-Require-Preflight: true' | jq -e '. == {"data":{"__typename":"Query"}}' > /dev/null; then
+            echo "Apollo server is up and healthy"
+          else
+            echo "Apollo server is down"
+            exit 1
+          fi
+          echo "Stopping production server..."
+          kill -9 $PROD_PID || echo "Process already stopped."
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply .github/workflows/pull-request.yml:394]: not required

File: .github/workflows/pull-request.yml [hunk 4]
```diff
@@ -296,6 +273,133 @@ jobs:
...
+          sleep 10
+          if curl -f http://localhost:4000 | jq -e '. == {"talawa-version":"v1","status":"healthy"}' > /dev/null; then
+            echo "Development server is up and healthy"
+          else
+            echo "Development server is down"
+            kill $DEV_PID
+            exit 1
+          fi
+          kill $DEV_PID
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply .github/workflows/pull-request.yml:355]: not really required

**Comment by adithyanotfound** [comment]
The build currently only works with node version 21 and below. This is because some packages are still using CommonJs for imports. If we update the outdated packages this should get fixed and we can use node v22.

**Comment by palisadoes** [comment]
1. Please update the packages
2. Also update the `.node-version` file

**Comment by palisadoes** [comment]
Please make sure coderabbit.ai approves your changes.

File: .github/workflows/pull-request.yml [hunk 5]
```diff
@@ -296,6 +273,145 @@ jobs:
...
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Generate Secrets
+        run: |
+          echo "Generating secrets..."
+          echo "ACCESS_TOKEN_SECRET=$(openssl rand -hex 32)" >> $GITHUB_ENV
+          echo "REFRESH_TOKEN_SECRET=$(openssl rand -hex 32)" >> $GITHUB_ENV
+          echo "SERVER_PORT=4000" >> $GITHUB_ENV
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply .github/workflows/pull-request.yml]: done

File: tests/resolvers/Query/agendaItemByEvent.spec.ts [hunk 6]
```diff
@@ -8,7 +8,7 @@
 import type { TestEventType } from "../../helpers/events";
 
 let MONGOOSE_INSTANCE: typeof mongoose;
-let testEvent: TestEventType;
+const testEvent: TestEventType = null as any;
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply tests/resolvers/Query/agendaItemByEvent.spec.ts]: done

File: tests/resolvers/Query/fundsByOrganization.spec.ts [hunk 7]
```diff
@@ -8,7 +8,7 @@
 import type { TestOrganizationType } from "../../helpers/userAndOrg";
 
 let MONGOOSE_INSTANCE: typeof mongoose;
-let testOrganization: TestOrganizationType;
+const testOrganization: TestOrganizationType = null as any;
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply tests/resolvers/Query/fundsByOrganization.spec.ts]: done

File: tests/resolvers/Query/agendaItemByOrganization.spec.ts [hunk 8]
```diff
@@ -8,7 +8,7 @@
 import type { TestOrganizationType } from "../../helpers/userAndOrg";
 
 let MONGOOSE_INSTANCE: typeof mongoose;
-let testOrganization: TestOrganizationType;
+const testOrganization: TestOrganizationType = null as any;
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply tests/resolvers/Query/agendaItemByOrganization.spec.ts]: done

File: .github/workflows/pull-request.yml [hunk 9]
```diff
@@ -296,6 +273,147 @@ jobs:
...
+            exit 1
+          fi
+          echo "Stopping development server..."
+          kill -9 $DEV_PID || echo "Process already stopped."
+
+      - name: Generate SSL Certificates for Production
+        run: |
+          echo "Generating SSL certificates for production..."
+          npm run generate:ssl-private-key
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply .github/workflows/pull-request.yml:374]: it generates both

File: .github/workflows/pull-request.yml [hunk 10]
```diff
@@ -296,6 +273,147 @@
...
+        ports:
+          - 6379:6379
+    env:
+      MONGO_DB_URL: mongodb://localhost:27017/talawa-test-db
+      REDIS_HOST: localhost
+      REDIS_PORT: 6379
+      LAST_RESORT_SUPERADMIN_EMAIL: "abc@gmail.com"
+      COLORIZE_LOGS: "true"
+      LOG_LEVEL: "info"
+
```
[the remark this answers is not in the document]
Comment (adithyanotfound) [reply .github/workflows/pull-request.yml:309]: Not required for now.

**Review by palisadoes — CHANGES_REQUESTED** [review]
Please update the `.node-version` file. It is used by the `fnm` manager referenced in the INSTALLATION.md file

**Review by palisadoes — APPROVED** [review]

**Pull request merged**